arrays.qmd

```diff
@@ -5,105 +5,234 @@
 
-```
-update(S[]) <- S[i] - n_SI[i]
-update(I[]) <- I[i] + n_SI[i] - n_IR[i]
-update(R[]) <- R[i] + n_IR[i]
+```{r}
+sir_age <- odin({
+  # Equations for transitions between compartments by age group
+  update(S[]) <- S[i] - n_SI[i]
+  update(I[]) <- I[i] + n_SI[i] - n_IR[i]
+  update(R[]) <- R[i] + n_IR[i]
+  
+  # Individual probabilities of transition:
+  p_SI[] <- 1 - exp(-lambda[i] * dt) # S to I
+  p_IR <- 1 - exp(-gamma * dt) # I to R
+  
+  # Calculate force of infection
+  ## age-structured contact matrix: m[i, j] is mean number of contacts an
+  ## individual in group i has with an individual in group j per time unit
+  m <- parameter()
+  ## here s_ij[i, j] gives the mean number of contacts and individual in group
+  ## i will have with the currently infectious individuals of group j
+  s_ij[, ] <- m[i, j] * I[j]
+  ## lambda[i] is the total force of infection on an individual in group i 
+  lambda[] <- beta * sum(s_ij[i, ])
+  
+  # Draws from binomial distributions for numbers changing between
+  # compartments:
+  n_SI[] <- Binomial(S[i], p_SI[i])
+  n_IR[] <- Binomial(I[i], p_IR)
+  
+  initial(S[]) <- S0[i]
+  initial(I[]) <- I0[i]
+  initial(R[]) <- 0
+  
+  # User defined parameters - default in parentheses:
+  S0 <- parameter()
+  I0 <- parameter()
+  beta <- parameter(0.0165)
+  gamma <- parameter(0.1)
+  
+  # Dimensions of arrays
+  N_age <- parameter()
+  dim(S0) <- N_age
+  dim(I0) <- N_age
+  dim(S) <- N_age
+  dim(I) <- N_age
+  dim(R) <- N_age
+  dim(n_SI) <- N_age
+  dim(n_IR) <- N_age
+  dim(p_SI) <- N_age
+  dim(m) <- c(N_age, N_age)
+  dim(s_ij) <- c(N_age, N_age)
+  dim(lambda) <- N_age
+})
 ```
 
-The odin code
+In the odin code above
 
-```
+```r
 update(S[]) <- S[i] - n_SI[i]
 ```
 
 becomes (approximately)
 
-```
+```r
 for (int i = 0; i < S_length; ++i) {
   update_S[i] = S[i] + n_SI[i];
 }
 ```
 
-* Don't use index variables (`i`, `j`, `k`, etc) on the left hand side
-* Use multiple lines to account for boundary conditions
-* Can crash the program if out of bounds (we have plans for this)
+so there is an implicit indexing by `i` on the LHS in that equation of the odin code, and the generated code will then loop over all values of `i`. 
 
-## An age structured SIR model
 
+## Another example: an age-structured SIR model with vaccination
+Let's take the above model and additionally add some vaccination to it.
+
+```{r}
+sir_age_vax <- odin({
+  # Equations for transitions between compartments by age group
+  update(S[, ]) <- new_S[i, j]
+  update(I[, ]) <- I[i, j] + n_SI[i, j] - n_IR[i, j]
+  update(R[, ]) <- R[i, j] + n_IR[i, j]
+  
+  # Individual probabilities of transition:
+  p_SI[, ] <- 1 - exp(-rel_susceptibility[j] * lambda[i] * dt) # S to I
+  p_IR <- 1 - exp(-gamma * dt) # I to R
+  p_vax[, ] <- 1 - exp(-eta[i, j] * dt)
+  
+  # Force of infection
+  m <- parameter() # age-structured contact matrix
+  s_ij[, ] <- m[i, j] * sum(I[j, ])
+  lambda[] <- beta * sum(s_ij[i, ])
```

Comment:
We should spell out in the text that this is a matrix multiplication, and give the appropriate R code (`lambda <- beta * (m %*% I)` I think) and say that we're looking to simplify this in future?

Reply:
Added a note about this!